python-package/xgboost/_typing.py

```diff
@@ -0,0 +1,62 @@
+"""Shared typing definition."""
+import ctypes
+import os
+from typing import Optional, List, Any, TypeVar, Union
+
+import numpy as np
+
+# os.PathLike/string/numpy.array/scipy.sparse/pd.DataFrame/dt.Frame/
+# cudf.DataFrame/cupy.array/dlpack
+DataType = Any
+
+# xgboost accepts some other possible types in practice due to historical reason, which is
+# lesser tested.  For now we encourage users to pass a simple list of string.
+FeatureNames = Optional[List[str]]
+
+ArrayLike = Any
+PathLike = Union[str, os.PathLike]
+CupyT = ArrayLike  # maybe need a stub for cupy arrays
+NdarrayOrCupyT = Union[np.ndarray, CupyT]
```

Comment:
I think we might be better off to return `Any` for prediction.  The return type is actually a dependent type, which means given the input type, the return type is known so it's not a union.  From a static type language's point of view (like c++, rust), the union can be changed at runtime, but the return type is determined at compile time.

Reply:
mypy doesn't support dependent type yet, you can find some related feature requests in their issue list.  To provide a concrete example, if the users use a cupy array as input to predict, then they know the output is a cupy array.  However, using `Union` the users will be asked to handle both numpy array and cupy array.

Reply:
I agree. By the way, should we use string literal for cupy types, e.g. `"cupy.ndarray"` instead of `CupyT` here https://github.com/bridgream/xgboost/blob/mypy/python-package/xgboost/core.py#L276

Reply:
@bridgream I think it will show an error as mypy couldn't find cupy on the test environment (cupy is an optional dependency)

Reply:
On a second thought, overloading might help, but the change might be tedious as cupy is an optional dependency.

Reply:
Maybe we should follow your solution in [this PR](https://github.com/dmlc/xgboost/pull/7242/files#diff-9748ea5ca0c5306e9992e33ff13359e683c4b9dc62b5a291b35b2966899aca85) by creating an interface. Since xgboost only touches some basic interface, this shouldn't take too much work. We could then introduce `@overload` to achieve better hints. 
I can try that next.